src/types.rs

```diff
@@ -825,6 +825,169 @@ impl SourceMap {
+        // both start and end.
+        // We want to compare `left` tokens and `right` by line/column numbers in the "middle" file.
+        // These line/column numbers are the `src_line/col` for `left` tokens and `dst_line/col` for
+        // the right tokens.
+        let mut left_tokens = left.tokens.clone();
+        left_tokens.sort_unstable_by_key(|t| (t.src_line, t.src_col));
+        let mut left_token_iter = left_tokens.iter().peekable();
+        let mut left_ranges = Vec::new();
+
+        while let Some(&t) = left_token_iter.next() {
+            let (end_line, end_col) = left_token_iter
+                .peek()
+                .map_or((u32::MAX, u32::MAX), |&&t| (t.src_line, t.src_col));
+            left_ranges.push(Range {
+                start: (t.src_line, t.src_col),
+                end: (end_line, end_col),
+                value: t,
+            });
+        }
+
+        let mut right_tokens = right.tokens.clone();
+        right_tokens.sort_unstable_by_key(|t| (t.dst_line, t.dst_col));
+        let mut right_token_iter = right_tokens.iter().peekable();
+        let mut right_ranges = Vec::new();
+
+        while let Some(&t) = right_token_iter.next() {
+            let (end_line, end_col) = right_token_iter
+                .peek()
+                .map_or((u32::MAX, u32::MAX), |&&t| (t.dst_line, t.dst_col));
+            right_ranges.push(Range {
+                start: (t.dst_line, t.dst_col),
+                end: (end_line, end_col),
+                value: t,
+            });
+        }
+
+        let mut right_ranges_iter = right_ranges.iter_mut();
+
+        let Some(mut right_range) = right_ranges_iter.next() else {
+            return builder.into_sourcemap();
+        };
+
+        // Iterate over `left.ranges` (sorted by `src_line/col`). For each such range, consider
+        // all `right.ranges` which overlap with it.
+        for &left_range in &left_ranges {
+            // The `left_range` offsets lines and columns by a certain amount. All `right_ranges`
+            // it covers will get the same offset.
+            let (line_diff, col_diff) = (
+                left_range.value.dst_line as i32 - left_range.value.src_line as i32,
+                left_range.value.dst_col as i32 - left_range.value.src_col as i32,
+            );
+
+            // Skip `right_ranges` that are entirely before the `left_range`.
+            while right_range.end <= left_range.start {
+                match right_ranges_iter.next() {
+                    Some(r) => right_range = r,
+                    None => return builder.into_sourcemap(),
+                }
+            }
+
+            // At this point `right_range.end` > `left_range.start`
+
+            // If the first `right_range` starts after the `left_range`,
+            // there's a gap between the `left_range` and the `right_range`.
+            // Add an "empty" mapping for that gap.
```

Comment:
isn’t this always the case because you skipped all the `right_range` that start/end before?

I also noticed this "empty" mapping in the tests. If you are composing with the "identity" sourcemap, there shouldn’t be any kind of bogus mapping inserted at the start.

Reply:
Well, `right_range` might start *exactly* at `left_range`, in which case this wouldn't happen. I've been going back and forth about this; you're probably right that it's cleaner without this special case. The identity composition in particular makes this seem like a flaw.